### PR TITLE
Issue #121

### PR DESCRIPTION
**Pull request by fadlinurhasan**

### CODE REVIEW

File: app/models/infrastructure.rb [hunk 1]
```diff
@@ -121,4 +129,19 @@ def allow_delete?
       'DELETE_ERROR'
     ].include?(self.provisioning_status) && self.status == 'INACTIVE'
   end
+
```
Comment (giosakti) [inline app/models/infrastructure.rb]: what are these methods, do you have tests for this?

**Comment by giosakti** [comment]
cc @tarabaskara see my reviews above

File: db/migrate/20190220114400_create_component_templates.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class CreateComponentTemplates < ActiveRecord::Migration[5.2]
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/migrate/20190220114400_create_component_templates.rb]: done

File: db/migrate/20190220114000_create_component_properties.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class CreateComponentProperties < ActiveRecord::Migration[5.2]
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/migrate/20190220114000_create_component_properties.rb]: done https://github.com/BaritoLog/BaritoMarket/commit/16492edeb269ce84597a96641048762640db6fc1

File: app/models/infrastructure.rb (hunk 1 above)
Comment (fadlinurhasan) [reply app/models/infrastructure.rb]: I believe these methods were there before we made this PR.

File: app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 module ChefHelper
   class BaritoFlowConsumerRoleAttributesGenerator < GenericRoleAttributesGenerator
     def initialize(component, infrastructure_components, opts = {})
-      @kafka_hosts = fetch_hosts_address_by(
+      kafka_hosts = fetch_hosts_address_by(
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb]: done https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb [hunk 5]
```diff
@@ -11,42 +11,42 @@ def initialize(component, infrastructure_components, opts = {})
...
-          }
-        },
-        'run_list' => ["role[#{@role_name}]"]
+
+      return {} if @consumer_attrs.nil?
+      return update_attrs
+    end
+
+    def update_attrs
+      @consumer_attrs['barito-flow']['consumer']['version'] = 'v0.11.8'
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb [hunk 6]
```diff
@@ -1,47 +1,43 @@
 module ChefHelper
   class BaritoFlowProducerRoleAttributesGenerator < GenericRoleAttributesGenerator
     def initialize(component, infrastructure_components, opts = {})
-      @kafka_hosts = fetch_hosts_address_by(
+      kafka_hosts = fetch_hosts_address_by(
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb [hunk 7]
```diff
@@ -1,47 +1,43 @@
...
-            'consul.json' => {
-              'bind_addr' => @ipaddress
-            }
-          }
-        },
-        'run_list' => ["role[#{@role_name}]"]
+    def update_attrs
+      @producer_attrs['barito-flow']['producer']['version'] = 'v0.11.8'
+      @producer_attrs['barito-flow']['producer']['env_vars'] = {
+        'BARITO_PRODUCER_ADDRESS'     => ':8080',
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/elasticsearch_role_attributes_generator.rb [hunk 8]
```diff
@@ -15,47 +15,35 @@ def initialize(component, infrastructure_components, opts = {})
...
-          }
-        },
-        'run_list' => ["role[#{@role_name}]", 'recipe[elasticsearch_wrapper_cookbook::elasticsearch_set_replica]']
-      }
+      return {} if @elastic_attrs.nil?
+      return update_attrs
+    end
+
+    def update_attrs
+      @elastic_attrs['elasticsearch']['version'] = '6.3.0'
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/elasticsearch_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/kafka_role_attributes_generator.rb [hunk 9]
```diff
@@ -11,50 +11,32 @@ def initialize(component, infrastructure_components, opts = {})
...
-              }
-            ]
-          }
-        }
-        attrs['run_list'] << 'recipe[datadog::default]'
-        attrs['run_list'] << 'recipe[datadog::kafka_datadog]'
+        @kafka_attrs['datadog']['datadog_api_key'] = Figaro.env.datadog_api_key
+        @kafka_attrs['datadog']['datadog_hostname'] = @hostname
+        @kafka_attrs['datadog']['kafka']['instances'][0]['cluster_name'] = @cluster_name
+        @kafka_attrs['run_list'] << 'recipe[datadog::default]'
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/kafka_role_attributes_generator.rb]: We remove runlist value in template and set it on generator process. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/kibana_role_attributes_generator.rb [hunk 10]
```diff
@@ -9,30 +9,26 @@ def initialize(component, infrastructure_components, opts = {})
...
     end
 
     def generate
-      elasticsearch_url = "http://#{@elasticsearch_hosts.first}:#{@elasticsearch_port}"
+      return {} if @kibana_attrs.nil?
+      return update_attrs
+    end
+
+    def update_attrs
+      @kibana_attrs['kibana']['version'] = '6.3.0'
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/kibana_role_attributes_generator.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/zookeeper_role_attributes_generator.rb [hunk 11]
```diff
@@ -10,46 +10,32 @@ def initialize(component, infrastructure_components, opts = {})
...
-          }
-        },
-        'run_list' => ["role[#{@role_name}]"]
-      }
+      return {} if @zookeeper_attrs.nil?
+      return update_attrs
+    end
+
+    def update_attrs
+      @zookeeper_attrs['zookeeper']['hosts'] = @hosts
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/zookeeper_role_attributes_generator.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: db/seeds/master_data/component_property.rb [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    },
+    "datadog":{
+      "zk":{
+        "instances":[{"host":"localhost", "port":2181, "tags":[], "cluster_name":""}]
+      },
+      "datadog_api_key":"",
+      "datadog_hostname":""
+    },
+    "run_list":["role[zookeeper]", "recipe[datadog::default]", "recipe[datadog::zk_datadog]"],
+    "zookeeper":{"hosts":[""],"my_id":1}
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/component_property.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: db/seeds/master_data/component_property.rb [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    }, 
+    "run_list":[
+      "role[elasticsearch]", 
+      "recipe[elasticsearch_wrapper_cookbook::elasticsearch_set_replica]", 
+      "recipe[datadog::default]", 
+      "recipe[datadog::elastic_datadog]"
+    ], 
+    "elasticsearch":{
+      "version":"6.3.0", 
+      "cluster_name":"guja",
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/component_property.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/121300ac01a8ac4afb71226d995be3fa6b6df60c

File: db/seeds/master_data/component_template.rb [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+# Seeds
+# -----------------------------------------------------------------------------
+
+### Staging
+ComponentTemplate.create!(
+  env: "staging",
+  name: "Staging - Small",
+  max_tps: 100,
+  instances: {
+    "yggdrasil": {"count": 0, "seq": 0},
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/component_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/ad617ccff2e25d878da5926e4bdb6cef069af8b9

File: app/models/component_template.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+class ComponentTemplate < ApplicationRecord
+  attr_accessor :infrastructure
+
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/models/component_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/ad617ccff2e25d878da5926e4bdb6cef069af8b9

File: app/helpers/app_groups_helper.rb [hunk 16]
```diff
@@ -1,6 +1,6 @@
 module AppGroupsHelper
   def max_tps(infrastructure)
-    TPS_CONFIG[infrastructure.capacity]['max_tps']
+    infrastructure.component_template.try(:max_tps) || TPS_CONFIG[infrastructure.capacity]['max_tps']
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/helpers/app_groups_helper.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/fc3c1364dcd46a12b07047a38967f4d1907f2028

File: app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb [hunk 17]
```diff
@@ -1,47 +1,43 @@
...
   class BaritoFlowProducerRoleAttributesGenerator < GenericRoleAttributesGenerator
     def initialize(component, infrastructure_components, opts = {})
-      @kafka_hosts = fetch_hosts_address_by(
+      kafka_hosts = fetch_hosts_address_by(
         infrastructure_components, 'category', 'kafka')
       @kafka_port = opts[:kafka_port] || 9092
       @consul_hosts = fetch_hosts_address_by(
         infrastructure_components, 'category', 'consul')
-      @max_tps = TPS_CONFIG[component.infrastructure.capacity]['max_tps'] || 10
+      @max_tps = component.infrastructure.component_template.try(:max_tps) || TPS_CONFIG[component.infrastructure.capacity]['max_tps'] || 10
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/fc3c1364dcd46a12b07047a38967f4d1907f2028

File: app/models/app_group.rb [hunk 18]
```diff
@@ -46,6 +46,6 @@ def available?
   end
 
   def max_tps
-    TPS_CONFIG[self.infrastructure.capacity]['max_tps']
+    self.infrastructure.cluster_template.try(:max_tps)
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/models/app_group.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/31cc9c578d54d46135ffc88561ce9bd5cc748711

File: app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb [hunk 19]
```diff
@@ -1,47 +1,36 @@
...
   class BaritoFlowProducerRoleAttributesGenerator < GenericRoleAttributesGenerator
     def initialize(component, infrastructure_components, opts = {})
-      @kafka_hosts = fetch_hosts_address_by(
+      kafka_hosts = fetch_hosts_address_by(
         infrastructure_components, 'category', 'kafka')
       @kafka_port = opts[:kafka_port] || 9092
       @consul_hosts = fetch_hosts_address_by(
         infrastructure_components, 'category', 'consul')
-      @max_tps = TPS_CONFIG[component.infrastructure.capacity]['max_tps'] || 10
+      @max_tps = component.infrastructure.cluster_template.try(:max_tps) || 10
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_producer_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/31cc9c578d54d46135ffc88561ce9bd5cc748711

File: app/helpers/app_groups_helper.rb [hunk 20]
```diff
@@ -1,6 +1,6 @@
 module AppGroupsHelper
   def max_tps(infrastructure)
-    TPS_CONFIG[infrastructure.capacity]['max_tps']
+    infrastructure.cluster_template.try(:max_tps)
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/helpers/app_groups_helper.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/31cc9c578d54d46135ffc88561ce9bd5cc748711

File: app/controllers/api/v2/apps_controller.rb [hunk 21]
```diff
@@ -157,8 +157,8 @@ def generate_profile_response(app)
         },
         kafka: {
           topic_name: app.topic_name,
-          partition: TPS_CONFIG[infrastructure.capacity]['kafka_options']['partition'],
-          replication_factor: TPS_CONFIG[infrastructure.capacity]['kafka_options']['replication_factor'],
+          partition: infrastructure.cluster_template.kafka_options['partition'],
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/controllers/api/v2/apps_controller.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/31cc9c578d54d46135ffc88561ce9bd5cc748711

File: app/controllers/api/apps_controller.rb [hunk 22]
```diff
@@ -159,8 +159,8 @@ def generate_profile_response(app)
         },
         kafka: {
           topic_name: app.topic_name,
-          partition: TPS_CONFIG[infrastructure.capacity]['kafka_options']['partition'],
-          replication_factor: TPS_CONFIG[infrastructure.capacity]['kafka_options']['replication_factor'],
+          partition: infrastructure.cluster_template.kafka_options['partition'],
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/controllers/api/apps_controller.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/31cc9c578d54d46135ffc88561ce9bd5cc748711

File: app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/lib/chef_helper/barito_flow_consumer_role_attributes_generator.rb]: done. https://github.com/BaritoLog/BaritoMarket/commit/56c927c0ea5d17d24a23fb568eb3e6192a3d2b17

File: db/seeds/master_data/cluster_template.rb [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    {
+      "name": "barito-flow-consumer",
+      "count": 10
+    },
+    {
+      "name": "kibana",
+      "count": 1
+    }
+  ],
+  options: {"partition": 30, "replication_factor": 3, "max_tps": 1000}
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/cluster_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/71092e14b79be2167adc8f3712b3e399303fa419

File: spec/factories/cluster_templates.rb [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+                          "name": "barito-flow-consumer",
+                          "count": 1
+                        },
+                        {
+                          "name": "kibana",
+                          "count": 1
+                        }
+                      ]
+                    }
+    options   { {"partition": 1, "replication_factor": 1, "max_tps": 100} }
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply spec/factories/cluster_templates.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/71092e14b79be2167adc8f3712b3e399303fa419

File: spec/factories/infrastructures.rb [hunk 25]
```diff
@@ -6,6 +6,44 @@
     provisioning_status   Infrastructure.provisioning_statuses[:pending]
     status                Infrastructure.statuses[:inactive]
     consul_host           Faker::Internet.domain_name
+    options               { {"partition": 1, "replication_factor": 1, "max_tps": 100} }
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply spec/factories/infrastructures.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/71092e14b79be2167adc8f3712b3e399303fa419

File: db/seeds/master_data/cluster_template.rb [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+
+# Seeds
+# -----------------------------------------------------------------------------
+
+### Staging
+ClusterTemplate.create!(
+  name: "Small",
+  instances: [
+    {
+      "name": "yggdrasil",
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/cluster_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/c3a2e68ee3e33e520d1f9b5b8f255335cd7f6019

File: db/migrate/20190220114400_create_cluster_templates.rb [hunk 27]
```diff
@@ -0,0 +1,11 @@
+class CreateClusterTemplates < ActiveRecord::Migration[5.2]
+  def change
+    create_table :cluster_templates do |t|
+      t.string :name
+      t.jsonb :instances
+      t.jsonb :options
+      t.timestamps null: false
+    end
+    add_index :cluster_templates, :name
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/migrate/20190220114400_create_cluster_templates.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/ea953e79418e0d16660032f72c048ade9c1670a6

File: db/seeds/master_data/component_template.rb [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+        "version":"v0.11.8", 
+        "env_vars":{
+          "BARITO_CONSUL_URL":"",
+          "BARITO_CONSUL_KAFKA_NAME":"kafka", 
+          "BARITO_CONSUL_ELASTICSEARCH_NAME":"elasticsearch",
+          "BARITO_KAFKA_BROKERS":"", 
+          "BARITO_KAFKA_GROUP_ID":"barito-group", 
+          "BARITO_KAFKA_CONSUMER_TOPICS":"barito-log", 
+          "BARITO_ELASTICSEARCH_URL":"", 
+          "BARITO_PUSH_METRIC_URL":"https://barito.golabs.io/api/increase_log_count"
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/component_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/026f2c3f223d14f9e068eb1a9c4c125419bda797

File: spec/lib/chef_helper/barito_flow_consumer_role_attributes_generator_spec.rb [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+              "version":"v0.11.8", 
+              "env_vars":{
+                "BARITO_CONSUL_URL":"",
+                "BARITO_CONSUL_KAFKA_NAME":"kafka", 
+                "BARITO_CONSUL_ELASTICSEARCH_NAME":"elasticsearch",
+                "BARITO_KAFKA_BROKERS":"", 
+                "BARITO_KAFKA_GROUP_ID":"barito-group", 
+                "BARITO_KAFKA_CONSUMER_TOPICS":"barito-log", 
+                "BARITO_ELASTICSEARCH_URL":"", 
+                "BARITO_PUSH_METRIC_URL":"https://barito.golabs.io/api/increase_log_count"
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply spec/lib/chef_helper/barito_flow_consumer_role_attributes_generator_spec.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/026f2c3f223d14f9e068eb1a9c4c125419bda797

File: spec/services/barito_blueprint/bootstrapper_spec.rb [hunk 30]
```diff
@@ -86,7 +86,7 @@ module BaritoBlueprint
     end
 
     describe '#generate_bootstrap_attributes' do
-      it 'should return proper attributes based on the component category' do
+      it 'should return proper consul attributes based on the component category' do
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply spec/services/barito_blueprint/bootstrapper_spec.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/026f2c3f223d14f9e068eb1a9c4c125419bda797

File: app/models/cluster_template.rb [hunk 31]
```diff
@@ -0,0 +1,10 @@
+class ClusterTemplate < ApplicationRecord
+  attr_accessor :infrastructure
+
+  validates :name, :instances, :options, presence: true
+
+  has_many :infrastructures
+end
+
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply app/models/cluster_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/ea953e79418e0d16660032f72c048ade9c1670a6

File: db/seeds/master_data/cluster_template.rb [hunk 32]
```diff
@@ -0,0 +1,124 @@
+
+# Fetch dependency
+# -----------------------------------------------------------------------------
+
+# Seeds
+# -----------------------------------------------------------------------------
+
+### Staging
+ClusterTemplate.create!(
+  name: "Small",
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply db/seeds/master_data/cluster_template.rb]: Done. https://github.com/BaritoLog/BaritoMarket/commit/2abf0d6f47bb2d57543ba0d0fb89468d0f92225c

**Pull request merged**